app/test/vesting_test.go

```diff
@@ -0,0 +1,542 @@
+	vestingtypes "github.com/cosmos/cosmos-sdk/x/auth/vesting/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	distributiontypes "github.com/cosmos/cosmos-sdk/x/distribution/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+	tmtime "github.com/tendermint/tendermint/types/time"
+	"google.golang.org/grpc"
+)
+
+func TestVestingModule(t *testing.T) {
+	suite.Run(t, new(VestingModuleTestSuite))
+}
+
+type VestingModuleTestSuite struct {
+	suite.Suite
+
+	// Regular accounts
+	accounts       []string
+	mut            sync.Mutex
+	accountCounter int
+}
+
+func (s *VestingModuleTestSuite) SetupSuite() {
+	t := s.T()
+	t.Log("setting up vesting module test suite")
+	s.accounts = testfactory.GenerateAccounts(300)
+}
+
+func (s *VestingModuleTestSuite) unusedAccount() string {
+	s.mut.Lock()
+	acc := s.accounts[s.accountCounter]
+	s.accountCounter++
+	s.mut.Unlock()
+	return acc
+}
+
+func (s *VestingModuleTestSuite) NewNetworkWithGenesisOpt(kr keyring.Keyring, genesisOpts ...testnode.GenesisOption) (cctx testnode.Context, rpcAddr, grpcAddr string) {
+	tmCfg := testnode.DefaultTendermintConfig()
+	tmCfg.Consensus.TargetHeightDuration = time.Millisecond * 1
+	appConf := testnode.DefaultAppConfig()
+
+	for _, name := range s.accounts {
+		testfactory.NewBaseAccount(kr, name)
+	}
+
+	cctx, rpcAddr, grpcAddr = testnode.NewNetwork(s.T(), testnode.DefaultParams(), tmCfg, appConf, []string{}, genesisOpts...)
+	cctx.Keyring = kr
+
+	return cctx, rpcAddr, grpcAddr
+}
+
+func (s *VestingModuleTestSuite) TestGenesisDelayedVestingAccounts() {
+	// initial unlocked allocation to pay the gas fees
+	initBalanceForGasFee := sdk.NewCoin(app.BondDenom, sdk.NewInt(10))
+
+	tests := []struct {
+		name    string
+		endTime time.Time
+	}{
+		{
+			name:    "ValidVestingAccount",
+			endTime: time.Now().Add(time.Hour * 50),
+		},
+		{
+			name:    "ImmediateVesting",
+			endTime: time.Now(),
+		},
+		{
+			name:    "NegativeEndTime",
+			endTime: time.Now().Add(-time.Second * 10),
+		},
+	}
+
+	kr := testfactory.GenerateKeyring()
+
+	gsOpt := func(gs map[string]json.RawMessage) map[string]json.RawMessage {
+		vAccs := authtypes.GenesisAccounts{}
+		vBals := []banktypes.Balance{}
+		for i, tt := range tests {
+			tests[i].accName = fmt.Sprintf("vesting_%d", i)
+			bacc, coins := testfactory.NewBaseAccount(kr, tests[i].accName)
+			acc := vestingtypes.NewDelayedVestingAccount(bacc, coins, tt.endTime.Unix())
+			vAccs = append(vAccs, acc)
+			vBals = append(vBals, banktypes.Balance{
+				Address: acc.GetAddress().String(),
+				Coins:   acc.GetOriginalVesting().Add(initBalanceForGasFee),
+			})
+		}
+
+		var err error
+		gs, err = AddAccountsToGenesisState(gs, vAccs...)
+		assert.NoError(s.T(), err)
+
+		gs, err = AddBalancesToGenesisState(gs, vBals...)
+		assert.NoError(s.T(), err)
+
+		return gs
+	}
+
+	cctx, _, _ := s.NewNetworkWithGenesisOpt(kr, gsOpt)
```

Comment:
[blocking]
minor, but this should be run during the setup of the suite instead of per test

Reply:
Done.